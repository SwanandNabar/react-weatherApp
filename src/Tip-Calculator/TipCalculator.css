.tip-calculator {
    text-align: center;
    font-family: 'Pathway Gothic One', sans-serif;
    color: #272822;
    max-width: 400px;
    margin: 0 auto;
    padding: 60px 0;
}

.tip-form {
    background-color: #dddddd;
    padding: 40px;
    border-radius: 7px;
}

.tip-calculator h1 {
    font-size: 36px;
    margin: 0;
    padding: 0;
    padding-bottom: 30px;
    line-height: 1.5;
    color: #484848;
}


#calculatorForm .bill-without-tip, .drop-down, .party-size {
    box-sizing: border-box;
    outline: none;
    border: 3px solid #e6e6e6;
    border-radius: 5px;
    background-color: #fff;
    line-height: 60px;
    vertical-align: middle;
    height: 45px;
    color: #A8A8A8;
    font-size: 15px;
    width: 100%;
    padding: 0 15px;
    position: relative;
}

.btn-calculate {
    font-family: sans-serif;
    position: relative;
    z-index: 100;
    text-transform: uppercase;
    background-color: #BBB;
    line-height: 1.1;
    font-size: 12px;
    border-radius: 3px;
    color: #fff;
    padding: 10px 40px;
    font-weight: bold;
    transition: all 200ms ease;
    cursor: pointer;
    height: 40px;
    display: inline-block;
    margin-left: 10px;
    margin-top: 13px;
    width: 100%;
}

.btn-clear {
    margin-left: 10px;
}



.field {
    width: 100%;
    position: relative;
    padding: 10px;
    color: #A8A8A8;
}

.title {
    display: none
}

.split {
    position: relative;
    text-align: left;
    padding: 0 13px;
    width: 100%;
    font-size: 15px;
    line-height: 1.5;
    color: #A8A8A8;
}

.checkbox {
    transform: scale(1.5);
    float: right;
}

.split-checkbox {
    float: left;
    width: 75%;
}

.total, .tip-amount {
    text-align: left;
    color: #A8A8A8;
    padding-left: 11px;
    width: 100%;
    font-size: 20px;
}

#tipsText {
    position: relative;
    float: right;
}

input {
    color: #A8A8A8;
}