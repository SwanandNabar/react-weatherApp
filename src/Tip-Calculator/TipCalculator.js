import React, { Component } from "react";
import Cleave from 'cleave.js/react';
import './TipCalculator.css';

export default class TipCalculator extends Component {
    constructor(props) {
        super(props);
        this.state = {
            billWithoutTip: '',
            partySize: '',
            value: 0,
            tipAmount: 0,
            totalBill: 0,
            splitBill: false,
            formSubmitted: false
        }
        this.baseState = this.state;
        this.onTipAmountChange = this.onTipAmountChange.bind(this);
        this.toggleSplitBill = this.toggleSplitBill.bind(this);
        this.calculateTip = this.calculateTip.bind(this);
    }


    onTipAmountChange(event) {
        const target = event.target;
        const value = target.value;
        const name = target.name;

        let selectorValue = target.name === "value" ? document.querySelector(".drop-down").value : this.state.value;

        switch (selectorValue) {
            case "0.15":
                document.querySelector(".tip-form").style.border = "solid 1px #00008B";
                break;
            case "0.18":
                document.querySelector(".tip-form").style.border = "solid 1px #00FFFF";
                break;
            case "0.2":
                document.querySelector(".tip-form").style.border = "solid 1px #32CD32";
                break;
            case "":
                document.querySelector(".tip-form").style.border = "none";
                break;
            default:
                break;
        }


        this.setState({
            [name]: value
        })

    }

    handleValidation(event) {
        if (!event.target.billWithoutTip.value || !event.target.value || !event.target.partySize.value) {
            this.setState({
                formSubmitted: true
            });
        }
    }

    toggleSplitBill() {
        this.setState({
            splitBill: !this.state.splitBill
        })
    }

    calculateTip(event) {
        event.preventDefault();

        this.handleValidation(event);

        var initialBill = parseFloat((this.state.billWithoutTip).replace(/,/g, ''));
        var partySizeRoundedOff = parseInt((this.state.partySize).replace(/,/g, ''), 10);

        if (this.state.splitBill) {
            var amountToSplit = initialBill / partySizeRoundedOff;
            var tipWhenSplittingBill = amountToSplit * this.state.value;
            var finalTotalWhenSplittingBill = (amountToSplit + tipWhenSplittingBill) * partySizeRoundedOff;

            // set state when splitting bill
            this.setState({
                tipAmount: tipWhenSplittingBill,
                totalBill: finalTotalWhenSplittingBill
            })
        } else {
            var tipWhenNotSplittingBill = initialBill * this.state.value;
            var finalTotalWhenNotSplittingBill = initialBill + tipWhenNotSplittingBill;

            //set state when not splitting the bill
            this.setState({
                tipAmount: tipWhenNotSplittingBill,
                totalBill: finalTotalWhenNotSplittingBill
            })
        }

    }

    render() {
        return (
            <div className="tip-calculator">
                <h1>Tip Calculator</h1>
                <div className="tip-form">
                    <form onSubmit={this.calculateTip} id="calculatorForm">
                        <div className="field">
                            <span className="title">Total Bill</span>
                            <Cleave name="billWithoutTip"
                                className="bill-without-tip"
                                placeholder="Total Bill"
                                onChange={this.onTipAmountChange}
                                value={this.state.billWithoutTip}
                                options={{ numeral: true, numeralPositiveOnly: true, numeralDecimalScale: 6 }} />
                        </div>
                        {/* validation error */}
                        {this.state.formSubmitted && !this.state.billWithoutTip &&
                            <span style={{ color: "red" }}>Please enter valid amount</span>
                        }
                        <div className="field">
                            <span className="title">Select a tip amount:</span>
                            <select name="value" value={this.state.value} onChange={this.onTipAmountChange} className="drop-down">
                                <option value="">Select a tip amount</option>
                                <option value="0.15">15 %</option>
                                <option value="0.18">18 %</option>
                                <option value="0.2">20 %</option>
                            </select>
                        </div>
                        {/* validation error */}
                        {this.state.formSubmitted && !this.state.value &&
                            <span style={{ color: "red" }}>Please select a tip amount</span>
                        }
                        <div className="field">
                            <span className="title">Party Size:</span>
                            <Cleave name="partySize"
                                className="party-size"
                                placeholder="Party Size"
                                onChange={this.onTipAmountChange}
                                value={this.state.partySize}
                                options={{ numeral: true, numeralPositiveOnly: true, stripLeadingZeroes: true, numeralDecimalScale: 0 }} />
                        </div>
                        {/* validation error */}
                        {this.state.formSubmitted && !this.state.partySize &&
                            <span style={{ color: "red" }}>Please enter the party size</span>
                        }
                        <div className="split">
                            <span className="split-checkbox">Split the bill between {this.state.partySize} people?</span>
                            <input name="splitBill"
                                className="checkbox"
                                type="checkbox"
                                checked={this.state.splitBill}
                                onChange={this.toggleSplitBill} />
                        </div>
                        <button type="submit" className="btn-calculate">Calculate tip</button>
                    </form>
                    <div className="tip-amount">
                        {this.state.splitBill &&
                            <p>
                                Tip for each person: <span id="tipsText">${(this.state.tipAmount).toFixed(2)}</span>
                            </p>
                        }
                        {!this.state.splitBill &&
                            <p>
                                Tip: <span id="tipsText">${(this.state.tipAmount).toFixed(2)}</span>
                            </p>
                        }
                    </div>
                    <div className="total">
                        Total: <span id="tipsText">${(this.state.totalBill).toFixed(2)}</span>
                    </div>
                </div>
            </div>
        )
    }

}