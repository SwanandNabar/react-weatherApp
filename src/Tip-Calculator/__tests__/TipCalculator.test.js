import TipCalculator from '../TipCalculator';

let tipCalculator;

describe("TipCalculator", () => {
    beforeEach(() => {
        tipCalculator = new TipCalculator();
    })
    it('Loads the component', () => {
        expect(tipCalculator).toMatchSnapshot();
    });
    it('calls the onTipAmountChange function', () => {
        const event = {
            preventDefault() { },
            value: 0.15,
            target: {
                billWithoutTip: '',
                partySize: '',
                value: 0.15,
            }
        };
        tipCalculator.onTipAmountChange(event);
        expect(tipCalculator).toMatchSnapshot();
    });

    it('calls the calculateTip function', () => {

        const event = {
            preventDefault() { },
            target: {
                initialBill: '',
                partySizeRoundedOff: '',
                billWithoutTip: ''
            }
        }
        tipCalculator.handleValidation(event);
        expect(tipCalculator).toMatchSnapshot();
        tipCalculator.calculateTip(event);
        expect(tipCalculator).toMatchSnapshot();
    });

    it('calls the toggleSplitBill function', () => {
        tipCalculator.toggleSplitBill();
        expect(tipCalculator).toMatchSnapshot();
    });
});