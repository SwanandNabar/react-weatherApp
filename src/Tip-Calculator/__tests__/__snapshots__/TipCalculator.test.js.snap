// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TipCalculator Loads the component 1`] = `
TipCalculator {
  "baseState": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "calculateTip": [Function],
  "context": undefined,
  "onTipAmountChange": [Function],
  "props": undefined,
  "refs": Object {},
  "state": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "toggleSplitBill": [Function],
  "updater": Object {
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
  },
}
`;

exports[`TipCalculator calls the calculateTip function 1`] = `
TipCalculator {
  "baseState": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "calculateTip": [Function],
  "context": undefined,
  "onTipAmountChange": [Function],
  "props": undefined,
  "refs": Object {},
  "state": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "toggleSplitBill": [Function],
  "updater": Object {
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
  },
}
`;

exports[`TipCalculator calls the calculateTip function 2`] = `
TipCalculator {
  "baseState": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "calculateTip": [Function],
  "context": undefined,
  "onTipAmountChange": [Function],
  "props": undefined,
  "refs": Object {},
  "state": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "toggleSplitBill": [Function],
  "updater": Object {
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
  },
}
`;

exports[`TipCalculator calls the onTipAmountChange function 1`] = `
TipCalculator {
  "baseState": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "calculateTip": [Function],
  "context": undefined,
  "onTipAmountChange": [Function],
  "props": undefined,
  "refs": Object {},
  "state": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "toggleSplitBill": [Function],
  "updater": Object {
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
  },
}
`;

exports[`TipCalculator calls the toggleSplitBill function 1`] = `
TipCalculator {
  "baseState": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "calculateTip": [Function],
  "context": undefined,
  "onTipAmountChange": [Function],
  "props": undefined,
  "refs": Object {},
  "state": Object {
    "billWithoutTip": "",
    "formSubmitted": false,
    "partySize": "",
    "splitBill": false,
    "tipAmount": 0,
    "totalBill": 0,
    "value": 0,
  },
  "toggleSplitBill": [Function],
  "updater": Object {
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
  },
}
`;
